src/Lower.cpp

```diff
@@ -440,6 +443,33 @@ void lower_impl(const vector<Function> &output_funcs,
         debug(1) << "Skipping GPU offload...\n";
     }
 
+    // TODO: This needs to happen before lowering parallel tasks, because global
+    // images used inside parallel loops are rewritten from loads from images to
+    // loads from closure parameters. Closure parameters are missing the Buffer<>
+    // object, which needs to be found by infer_arguments here. Running
+    // infer_arguments prior to lower_parallel_tasks is a hacky solution to this
+    // problem. It would be better if closures could directly reference globals
+    // so they don't add overhead to the closure.
+    vector<InferredArgument> inferred_args = infer_arguments(s, outputs);
+
+    std::vector<LoweredFunc> closure_implementations;
+    debug(1) << "Lowering Parallel Tasks...\n";
+    s = lower_parallel_tasks(s, closure_implementations, pipeline_name, t);
+    // Process any LoweredFunctions added by other passes. In practice, this
+    // will likely not work well enough due to ordering issues with
+    // closure generating passes and instead all such passes will need to
+    // be done at once.
+    for (size_t i = initial_lowered_function_count; i < result_module.functions().size(); i++) {
+        result_module.functions()[i].body =
+            lower_parallel_tasks(result_module.functions()[i].body, closure_implementations,
+                                 result_module.functions()[i].name, t);
+    }
+    for (auto &lowered_func : closure_implementations) {
```

Comment:
This is either a bug, or I just don't understand the code: `lower_parallel_tasks()` *replaces* the contents of `closure_implementations()`, rather than adding to it, so at this point, this list is just whatever it happened to be on the final call to `lower_parallel_tasks()` in the loop at line 462-466. Did we want to process the union of all of them? Or just the list from the first call to `lower_parallel_tasks()`?

Reply:
Was this resolved?